### ISSUE TITLE
`ENOENT` with custom output and ESM module in NPM monorepo (including Nextjs): `no such file or directory, open /.../schema.prisma...`

### ISSUE THREAD

**Issue by slimshreydy** [prisma/prisma#17687]
### Bug description

I've been [seeing](https://github.com/prisma/prisma/issues/12823) [a](https://github.com/prisma/prisma/issues/12921) [slew](https://github.com/prisma/prisma/issues/16117) [of](https://github.com/prisma/prisma/issues/12853) [issues](https://github.com/prisma/prisma/issues/13233) involving using Prisma-based local packages in an NPM monorepo with Nextjs, and I've encountered a case I could not solve using the existing solutions.

As a recap of the other issues, it seems when using a Prisma-based package (i.e. a package that ships with its own `prisma.schema`) in Nextjs, the resulting webpack bundles will look for `schema.prisma` relative to the directory of the bundle rather than the directory of the package `dist`. This results in errors like this:

```
Error: ENOENT: no such file or directory, open '/.../.next/cache/webpack/client-development/schema.prisma'
```

Now there have been some proposed solutions thus far, but the **big difference is my package is an ESM package whereas the ones in the previous issues have been CJS packages**. For example, one of the workarounds proposed [here](https://github.com/kn0ll/prisma-monorepo-output-bug/pull/1/files) says to modify the Next config to use webpack externals and `outputFileTracingRoot` like so:

```
module.exports = {
  output: 'standalone',
  experimental: {
    outputFileTracingRoot: path.join(__dirname, '../../'),
  },
  webpack: (config, { buildId, dev, isServer, defaultLoaders, webpack }) => {
    config.externals = [...config.externals, '<PACKAGE>']

    return config
  },
}
```

However, if `<PACKAGE>` is an ESM module, this means webpack never transpiles it to CJS, resulting in a `Error: require() of ES Module` error when you try using the package in a Nextjs app. Hence, this fix does not work for ESM-based packages.

### How to reproduce

I have created a repro repo here: https://github.com/slimshreydy/prisma-monorepo-esm-output-bug. See the README for instructions on local setup.

### Expected behavior

Importing a package that uses prisma in Nextjs should look for the `schema.prisma` file in the package's dist directory rather than the `.next/` folder.

### Prisma information

```prisma
generator client {
  provider = "prisma-client-js"
  output = "./client"
}

datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}

model User {
  id        Int      @id @default(autoincrement())
  email     String   @unique
  createdAt DateTime @default(now())
}
```

```ts
import { NextApiRequest, NextApiResponse } from "next";
import { prisma } from "@prisma-monorepo-esm-output-bug/package/dist";

export default async (req: NextApiRequest, res: NextApiResponse) =>
  res.json(await prisma.user.findMany());
```

### Environment & setup

- OS: macOS 10.15
- Database: PlanetScale
- Node.js version: v18.12.1

### Prisma Version

```
prisma                  : 4.9.0
@prisma/client          : 4.9.0
Current platform        : darwin
Query Engine (Node-API) : libquery-engine ceb5c99003b99c9ee2c1d2e618e359c14aef2ea5 (at ../node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli ceb5c99003b99c9ee2c1d2e618e359c14aef2ea5 (at ../node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core ceb5c99003b99c9ee2c1d2e618e359c14aef2ea5 (at ../node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt ceb5c99003b99c9ee2c1d2e618e359c14aef2ea5 (at ../node_modules/@prisma/engines/prisma-fmt-darwin)
Format Wasm             : @prisma/prisma-fmt-wasm 4.9.0-42.ceb5c99003b99c9ee2c1d2e618e359c14aef2ea5
Default Engines Hash    : ceb5c99003b99c9ee2c1d2e618e359c14aef2ea5
Studio                  : 0.479.0
```

**Comment by slimshreydy** [comment]
@janpio Tagging you since it seems you've been addressing most of these issues!

**Comment by dpromisel** [comment]
+1

**Comment by janpio** [comment]
Thank you for the awesome issue, and the excellent reproduction. 

My colleague @millsp knows more about the technical details and will be able to categorize this when he has a look hopefully. I think we had other transpile cases, but do not recall if we had a good workaround or not.

**Comment by millsp** [comment]
Hey @slimshreydy thanks for the thorough issue description. We really appreciate the effort! I just tried your reproduction and I have a few remarks. Right now, your reproduction does not fail for me, as it seems you already enabled the workaround and found how to circumvent your ts/esm issue. Could you please provide a branch with the repro that matches your description?

I have seen this problem with TypeScript a few days ago so I am working on a better workaround. We are building a webpack plugin that will fix that for you. In short, this plugin ensures that the necessary files are correctly copied for you but does not conflict with your other settings. You can try it now via https://www.npmjs.com/package/experimental-prisma-webpack-plugin

```ts
const { PrismaPlugin } = require('experimental-prisma-webpack-plugin')

module.exports = {
  output: 'standalone',
  webpack: (config, { isServer }) => {
    if (isServer) {
      config.plugins = [...config.plugins, new PrismaPlugin()]
    }

    return config
  },
}
```

**Comment by slimshreydy** [comment]
Hi @millsp! Were you trying the app locally (using `npm run dev -w web` to start the server in dev mode), or did you deploy it to Vercel? I was encountering issues with the former but the latter worked fine for me.

Regardless, I was able to add `experimental-prisma-webpack-plugin` to my Nextjs app and I am happy to report it works! Both the local and hosted versions of my nextjs work fine now. No more ENOENT errors! Thanks for your quick work here.

**Comment by millsp** [comment]
All good then,maybe I did something wrong on my end. That said, I had reproductions at hand that describe exactly what you said, so I think it is fine. I am really happy that it works for you! Thanks for reporting. We will look into ways to completely remove the workaround in the future.

**Comment by adrianbienias** [comment]
I have the same problem.
Here's the repo that reproduces the issue:
https://github.com/adrianbienias/nextjs-prisma-monorepo-issue

**Comment by millsp** [comment]
Hey @adrianbienias, have you tried this [workaround](https://github.com/prisma/prisma/issues/17687#issuecomment-1414826476), does it work for you?

**Comment by adrianbienias** [comment]
@millsp, I've tried, and it doesn't work.
Using `experimental-prisma-webpack-plugin` and changes in Next.js config from the workaround you linked, I'm getting the error: `Could not find mapping for model <model_name>`

**Comment by millsp** [comment]
Thanks for the feedback, I will take a look.

**Comment by Rafcin** [comment]
> I have the same problem. Here's the repo that reproduces the issue: https://github.com/adrianbienias/nextjs-prisma-monorepo-issue
> 
> I've added a workaround branch to the repo and explained it in the README.
> 
> Steps that worked for me as a temporary solution are quite simple:
> 
> 1. Set `output` in `schema.prisma` to shared monorepo `node_modules`
> 
> ```
> output = "../../node_modules/@prisma/client/<package_name>"
> ```
> 
> 2. Import Prisma clients from `@prisma/client/<package_name>`

Hi, I've been running into a similar issue, Prisma seems to throw this error or crash my NextJs app without error on Ubuntu, but on my mac, Prisma just works; it's odd. I was reading your workaround branch, and I just had a quick question, did you miss importing the packages into the nextjs app, or am I missing something?

**Comment by adrianbienias** [comment]
> I was reading your workaround branch, and I just had a quick question, did you miss importing the packages into the nextjs app, or am I missing something?

@Rafcin you're asking about missing dependencies of prisma packages in nextjs app?
If so, [npm workspaces](https://docs.npmjs.com/cli/v7/using-npm/workspaces) symlink packages defined in the root directory, so each package can access another without explicitly defining it as dependency.

Probably it's not the best practice if you plan to publish packages separately, but inside a monorepo it works well.

**Comment by Rafcin** [comment]
@adrianbienias I did not know I could do that. I always have it as a dependency. Good to know!

Um, I lied in my last post; it seems that this method is a bit wonky. I think, at this point, it's me. I created a Create T3 app and put it in my monorepo, and I created a separate package for Prisma as you did; however, when I try to utilize Prisma, I get errors like `Cannot read properties of undefined (reading 'findUnique') TypeError: Cannot read properties of undefined (reading 'findUnique')` or `Cannot read properties of undefined (reading 'account') TypeError: Cannot read properties of undefined (reading 'account')` when using next-auth. Prisma drives me a bit nuts.

**Comment by adrianbienias** [comment]
@Rafcin Packages with my repo packages aren't separated because of prisma, they are separate to represent a random package that uses it's own `schema.prisma`.

T3 app should work well the same way.

In `monorepo-root/t3-app/prisma/schema.prisma`
set for the client generator `output   = "../../node_modules/@prisma/client/t3-app"`

and in `db.ts`
set `import { PrismaClient } from "@prisma/client/t3-app";`

**Comment by Rafcin** [comment]
@adrianbienias I see, ok that makes sense.

So I spent the rest of the day tinkering with it, I was getting it to work but Nextjs would do a hard refresh, and just stop working anytime I invoked Prisma and it gave me no errors. I thought the different approaches weren't working until I purged my entire machine of nvm, node, NPM and Pnpm and now it works fine :/

Honestly for days I thought I was doing something wrong. Now that it works, I setup multiple schemas and put them in their own packages, and to avoid any issues I generated the client code in the package and just exported it, and that seems to work well!
